python/lsst/ap/pipe/ap_pipe.py

```diff
@@ -21,356 +21,327 @@
+            diffImResults = None
+        else:
+            diffImResults = self.runDiffIm(calexpRef, templateIds)
+
+        # No reasonable way to check if Association finished successfully
+        associationResults = self.runAssociation(calexpRef)
+
+        return pipeBase.Struct(
+            fullMetadata = self.getFullMetadata(),
+            l1Database = associationResults.l1Database,
+            ccdProcessor = processResults.taskResults if processResults else None,
+            differencer = diffImResults.taskResults if diffImResults else None,
+            associator = associationResults.taskResults if associationResults else None
+        )
+
+    @pipeBase.timeMethod
+    def runProcessCcd(self, sensorRef):
+        """Perform ISR with ingested images and calibrations via processCcd.
+
+        The output repository associated with ``sensorRef`` will be populated with the
+        usual post-ISR data (bkgd, calexp, icExp, icSrc, postISR).
+
+        Parameters
+        ----------
+        sensorRef : `lsst.daf.persistence.ButlerDataRef`
+            Data reference for raw data.
+
+        Returns
+        -------
+        fullMetadata : `lsst.daf.base.PropertySet`
+            The metadata produced by the Task. Intended as a transitional API
+            for ap_verify, and may be removed later.
+        taskResults : `lsst.pipe.base.Struct`
+            The output of the task assigned to ``config.ccdProcessor`
+            (by default, `lsst.pipe.tasks.ProcessCcdTask.run`)
+
+        Notes
+        -----
+        The input repository corresponding to ``sensorRef`` must already contain the refcats.
+        By default, the configuration for astrometric reference catalogs uses Gaia
+        and the configuration for photometry reference catalogs uses Pan-STARRS.
+        """
+        self.log.info("Running ProcessCcd...")
+        result = self.ccdProcessor.run(sensorRef)
+        return pipeBase.Struct(
+            fullMetadata = self.ccdProcessor.getFullMetadata(),
+            taskResults = result
+        )
+
+    @pipeBase.timeMethod
+    def runDiffIm(self, sensorRef, templateIds=None):
+        """Do difference imaging with a template and a science image
+
+        The output repository associated with ``sensorRef`` will be populated with difference images
+        and catalogs of detected sources (diaSrc, diffexp, and metadata files)
+
+        Parameters
+        ----------
+        sensorRef: `lsst.daf.persistence.ButlerDataRef`
+            Data reference for multiple dataset types, both input and output.
+        templateIds : `list` of `dict`, optional
+            A list of parsed data IDs for templates to use. Only used if
+            ``config.differencer.getTemplate`` is configured to do so.
+            Data IDs must contain only a visit. Specific implementations of
+            ``getTemplate`` may impose additional restrictions.
```

Comment:
This requirement is imposed by `ImageDifferenceTask`: https://github.com/lsst/pipe_tasks/blob/master/python/lsst/pipe/tasks/imageDifference.py#L792. They don't bother documenting this parameter at all in `run`, but I don't see why they'd impose it at the command-line parser level if the implementation didn't (potentially?) require it.

Reply:
This may be a reference to DM-11865: "ImageDifference should allow coadd templates and output to live in different repos"? Templates are treated rather like static data products in ap_pipe and that can create some interesting situations.

Reply:
This is the data ID used for calexp templates. I don't think DM-11865 is relevant.

Reply:
Interestingly, the visit-only rule isn't documented anywhere in `GetCalexpAsTemplateTask`, though it may be a de-factor limitation of their code.

Reply:
On second thought, this may all be moot, because it occurs to me that I'm using inconsistent levels of abstraction by assuming `ImageDifferenceTask` but not `GetCalexpAsTemplateTask`. If I replace the last two sentences with "`config.differencer` or `config.differencer.getTemplate` may restrict the allowed dataIds", would there be any problem, Russel?

Reply:
@kfindeisen I think the help text you linked to  is saying that only visit is used; other arguments are ignored. If you are suggesting replacing the text that begins "Data IDs must contain..." then yes that would be fine.